Cannot build Go encoder

`joshua@WOPR:~$ sudo apt install golang-go` # version 2:1.13~1ubuntu2
...
```console
joshua@WOPR:~$ git clone https://github.com/tonysavon/TSCrunch
Cloning into 'TSCrunch'...
...
joshua@WOPR:~$ cd TSCrunch/
joshua@WOPR:~/TSCrunch$ ls
LICENSE  README.md  TSCrunch.java  benchmark.png  bin  boot_code.h  decrunch.asm  decrunch_extreme.asm  decrunch_small.asm  readme.txt  tscrunch.c  tscrunch.go
joshua@WOPR:~/TSCrunch$ go build -o tscrunch tscrunch.go
# command-line-arguments
./tscrunch.go:185:15: undefined: os.ReadFile
./tscrunch.go:194:2: undefined: os.WriteFile
./tscrunch.go:200:25: cannot convert data[i:] (type []byte) to type *[3]byte
./tscrunch.go:320:25: cannot convert prefix (type []byte) to type *[3]byte
```

also the older build method that I used for an earlier version did not work:
```console
joshua@WOPR:~/TSCrunch$ mkdir -p bin/this
joshua@WOPR:~/TSCrunch$ go mod init tscrunch
go: creating new go.mod: module tscrunch
joshua@WOPR:~/TSCrunch$ go mod tidy
joshua@WOPR:~/TSCrunch$ go build -o bin/this/tscrunch
can't load package: package tscrunch: C source files not allowed when not using cgo or SWIG: tscrunch.c
```
tried again with 
`joshua@WOPR:~$ sudo apt install gccgo-go ` # version 2:1.13~1ubuntu2
...
```console
joshua@WOPR:~$ git clone https://github.com/tonysavon/TSCrunch
Cloning into 'TSCrunch'...
..
joshua@WOPR:~$ cd TSCrunch/
joshua@WOPR:~/TSCrunch$ go build -o tscrunch tscrunch.go
# command-line-arguments
./tscrunch.go:185:18: error: reference to undefined identifier ‘os.ReadFile’
  185 |  data, err := os.ReadFile(f)
      |                  ^
./tscrunch.go:194:5: error: reference to undefined identifier ‘os.WriteFile’
  194 |  os.WriteFile(f, data, 0666)
      |     ^
./tscrunch.go:113:44: error: shift count not unsigned integer
  113 |     specializedLiteralLengths |= uint32(1) << transition.size
      |                                            ^
./tscrunch.go:119:43: error: shift count not unsigned integer
  119 |    if specializedLiteralLengths&(uint32(1)<<size) != 0 {
      |                                           ^
./tscrunch.go:200:25: error: invalid type conversion
  200 |   key := *(*[MINLZ]byte)(data[i:])
      |                         ^
./tscrunch.go:320:25: error: invalid type conversion
  320 |   key := *(*[MINLZ]byte)(prefix)
      |                         ^
```

--- a/README.md
+++ b/README.md
@@ -66,7 +66,10 @@ On Linux and other Unix-like systems, link the math library and omit the `.exe` 
 Building the Go encoder
 =======================
 
+	# Go 1.13 or newer
 	go build -o tscrunch tscrunch.go
+
+Build `tscrunch.go` explicitly as shown above. Running `go build .` in the repository root also includes `tscrunch.c`, which requires cgo and is not needed for the Go encoder.
 
 Building the Java encoder
 =========================

--- a/readme.txt
+++ b/readme.txt
@@ -72,7 +72,10 @@ On Linux and other Unix-like systems, link the math library and omit the .exe su
 Building the Go encoder
 =======================
 
+	# Go 1.13 or newer
 	go build -o tscrunch tscrunch.go
+
+Build tscrunch.go explicitly as shown above. Running go build . in the repository root also includes tscrunch.c, which requires cgo and is not needed for the Go encoder.
 
 Building the Java encoder
 =========================

--- a/tscrunch.go
+++ b/tscrunch.go
@@ -8,6 +8,7 @@ import (
 	"bytes"
 	"flag"
 	"fmt"
+	"io/ioutil"
 	"math"
 	"os"
 	"runtime"
@@ -110,13 +111,13 @@ func shortestPathForward(transitions []token, starts []int, inputSize int) ([]to
 			destination := position + transition.size
 			relax(position, destination, tokenCost(position, destination, transition.tokentype), transition)
 			if transition.size <= LONGESTLITERAL {
-				specializedLiteralLengths |= uint32(1) << transition.size
+				specializedLiteralLengths |= uint32(1) << uint(transition.size)
 			}
 		}
 		literalLimit := min(LONGESTLITERAL, inputSize-position)
 		for size := 1; size <= literalLimit; size++ {
 			destination := position + size
-			if specializedLiteralLengths&(uint32(1)<<size) != 0 {
+			if specializedLiteralLengths&(uint32(1)<<uint(size)) != 0 {
 				continue
 			}
 			literal := LIT(position, size)
@@ -182,7 +183,7 @@ func max(x, y int) int {
 }
 
 func load_raw(f string) []byte {
-	data, err := os.ReadFile(f)
+	data, err := ioutil.ReadFile(f)
 	if err == nil {
 		return data
 	}
@@ -191,13 +192,14 @@ func load_raw(f string) []byte {
 }
 
 func save_raw(f string, data []byte) {
-	os.WriteFile(f, data, 0666)
+	ioutil.WriteFile(f, data, 0666)
 }
 
 func fillPrefixArray(data []byte, ctx *crunchCtx) {
 	ctx.prefixArray = make(map[[MINLZ]byte][]int)
 	for i := 0; i+MINLZ <= len(data); i++ {
-		key := *(*[MINLZ]byte)(data[i:])
+		var key [MINLZ]byte
+		copy(key[:], data[i:i+MINLZ])
 		ctx.prefixArray[key] = append(ctx.prefixArray[key], i)
 	}
 }
@@ -317,7 +319,8 @@ func findLZCandidates(src []byte, i int, minlz int, ctx *crunchCtx) lzCandidates
 	prefix := src[i : i+minlz]
 	x0 := max(0, i-LONGLZOFFSET)
 	if ctx.usePrefixArray {
-		key := *(*[MINLZ]byte)(prefix)
+		var key [MINLZ]byte
+		copy(key[:], prefix)
 		parray := ctx.prefixArray[key]
 		for o := sort.SearchInts(parray, i) - 1; o >= 0 && parray[o] >= x0; o-- {
 			j := parray[o]

